experimental/oak_baremetal_channel/SPEC.MD

```diff
@@ -0,0 +1,126 @@
+
+## Messages
+
+A message forms one half of an Invocation.
+
+### Request Message Definition
+
+A byte encoded request message MUST consist of the following seqeuential fields:
+
+- length, u32, little endian: length of the entire message, including this field
+- invocation_id, u32, little endian: Identifies the invocation the request
+  initiates
+- method_id, u32, little endian: identifies the method that is being invoked
+- padding, 4 bytes: reserved to maintain 8 byte alignment. MUST be unset and
+  ignored
+- body, variable length byte array: byte encoded method parameters
+
+### Response Message Definition
+
+A byte encoded response message MUST consist of the following seqeuential
+fields:
+
+- length, u32, little endian: length of the entire message, including this field
+- invocation_id, u32, little endian: MUST match the invocation_id of the
+  relevant Request Message
+- status_code, u32, little endian: response status as a
+  [gRPC status code](https://grpc.github.io/grpc/core/md_doc_statuscodes.html)
+- padding, 4 bytes: reserved to maintain 8 byte alignment. MUST be unset and
+  ignored
+- body, variable length byte array: For responses with an "Ok" status code the
+  body MUST be the byte encoded method return value. Else the body MUST be a
+  utf-8 encoded developer-facing error message.
+
+### Sending Messages
+
+To send a message, it MUST first be encoded into frames. The resulting frame set
+MUST be sent as an uninterrupted sequence.
+
+### Receiving Messages
+
+To receive a message the recipent incrementaly reads frames from the
+communication channel. The recipent MAY start parsing a message before all of
+its frames have been received.
+
+### Encoding Messages into Frames
+
+The byte representation of a message MUST be chunked. The last chunk MAY be
+shorter than 63,936 bytes (the maximum size of a frame body). Any preceeding
+chunks MUST be exactly 63,936 bytes in length.
+
+A frame MUST then be formed for each resulting message chunk, with the message
+chunk forming the frame body.
+
+Messages shorter than 63,936 bytes MUST be encoded as a single frame.
+
+### Decoding Frames into Messages
+
+Messages are parsed from Frames by appending the frame bodies in the order they
+are received.
+
+## Frames
+
+Messages are sent and received as frames. Each message can be fragmented into
+one or more frames.
+
+### Frame Definition
+
+A byte encoded frame MUST consist of the following seqeuential fields:
+
+- padding, four bytes: for future use
+- length, u16, little endian: Length of the entire frame, including this field.
+  The length of a frame MUST NOT exceed 64000 bytes
+- flags, two bytes: Bitflags, see [Frame Flags](#3-frame-flags-definition)
+- body, variable length byte array: Byte chunk of an encoded message. In order
+  to not exceed the maximum frame length, this byte array MUST NOT exceed 63,936
+  bytes
+
+### Frame Flags Definition
+
+- 0x0 (LSB): MUST be set for frames that are the first of the set that
```

Comment:
I always get this wrong (confusing little vs big endian, and LSB vs MSB); to confirm, is this expected to be the left-most bit in the message as serialized?

Reply:
Yes.

Reply:
Do you think we should use that language? MSB / LSB feels a bit cryptic :P

Reply:
AFAICT the numbering in the PR seems to be correct (or at least be matching fuchsia). Fuchsia labels 0 as the LSB (as do we), and appears to start using flags starting 0, with larger values unused (as do we). See https://fuchsia.dev/fuchsia-src/reference/fidl/language/wire-format#flags, Flags[0].

Reply:
Perhaps we don't need to define endianness of the bytes field, and we could just talk about the two bytes independently (`flags[0]` and `flags[1]`). But I don't think we can assume the same endianness across the rest of the spec. I mean, at some point we need to parse a u32 from bytes, and we need to know which endianness it is in order to call the appropriate Rust / C++ method, right? I don't think we would want to just blindly treat those bytes as an int of whatever endianness the machine happens to be running in.

Reply:
Fair, let's keep it then. I guess the downside is that if we do ever run on a processor that uses big endian natively we'd have to a lot of conversion. But seems unlikely. 

Interesting trivia fact (and what gave me the idea): Many common protocols used for actual networking between physical computers do not define endianness. UDP for one does not 😃